src/main/java/bio/terra/workspace/service/resource/controlled/flight/ValidateNoExistingGcsBucketStep.java

```diff
@@ -0,0 +1,50 @@
+package bio.terra.workspace.service.resource.controlled.flight;
+
+import bio.terra.cloudres.google.storage.BucketCow;
+import bio.terra.cloudres.google.storage.StorageCow;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.service.crl.CrlService;
+import bio.terra.workspace.service.resource.controlled.ControlledGcsBucketResource;
+import bio.terra.workspace.service.resource.exception.DuplicateResourceException;
+import bio.terra.workspace.service.workspace.WorkspaceService;
+
+/*
+ * Before starting a step to create a bucket, ensure that there's not already a bucket with
+ * the same name.
+ */
+public class ValidateNoExistingGcsBucketStep implements Step {
+
+  private final CrlService crlService;
+  private final ControlledGcsBucketResource resource;
+  private final WorkspaceService workspaceService;
+
+  public ValidateNoExistingGcsBucketStep(
+      CrlService crlService,
+      ControlledGcsBucketResource resource,
+      WorkspaceService workspaceService) {
+    this.crlService = crlService;
+    this.resource = resource;
+    this.workspaceService = workspaceService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    final String projectId = workspaceService.getRequiredGcpProject(resource.getWorkspaceId());
+    final StorageCow storageCow = crlService.createStorageCow(projectId);
+    final BucketCow existingBucket = storageCow.get(resource.getBucketName());
+    if (existingBucket != null) {
+      throw new DuplicateResourceException(
+          String.format("GCS Bucket with name %s already exists.", resource.getBucketName()));
```

Comment:
drive-by: Isn't there a race condition here where two create bucket flights can both get past this step and then try to create the same GcsBucket? I think we need to do the existence/creation as a transaction.
I was wondering if we should have the "cloud uri" column for controlled resources in the WM database and use that to ensure exclusive controlled resources.

Reply:
So doing it as a transaction would mean that we don't do it as two steps, right?

Reply:
We already know the bucket attributes, so we don't need to add anything in the database. We could do this:
Step 1: in a serialized DAO transaction, verify that no other controlled GCS bucket exists with that bucket name and, if not, write the metadata into the database. That prevents parallel bucket creates within WSM.
Step 2: check that the GCS bucket does not exist in GCP. This leaves a race between us and other bucket creators, but I don't see how to close that.
Step 3: create the bucket and just use it if already exists

Reply:
That sounds good to me.
I think Step 1 is generic for all controlled resources. Do we want to have a uniform cloud URI column, or write some switch based on the controlled resource type and look at the attributes json?

Reply:
Each controlled resource has a different comparison to do. The database column is jsonb, so we can construct SQL queries to do this, but the queries would be per-resource-type.

Reply:
Doing per-resource type queries sounds fine to me (what I was referring to "some switch based on the controlled resource type and look at the attributes json").
However, I'm not sure that I'm being clear in the alternative to that I'm suggesting: make a new text column "controlled_uri" in the resource table. Add a field to the `ControlledResource` class for the uri. As long as each controlled resource type can represent its underlying cloud resource uniqueness as a canonical string, we can put a uniqueness constraint on the "controlled_uri" text column and not write per-type SQL jsonb comparison switches. For buckets, the "controlled_uri" is the GCS bucket uri, e.g. `gs://my-bucket-id`. For BigQuery datasets, is the project id + dataset id. - e.g. 'my_project.my_dataset'. That does duplicate information in the attributes class.

I'm ok with either way, just want to consider both.

Reply:
Yeah, I like having the unique constraint live outside the JSON, so it becomes the database's job to check. That may be out of scope for this little ticket though.